### PR TITLE
[Merged by Bors] - refactor(CategoryTheory/Sites): define Functor.sheafPullback as a left adjoint

### PR DESCRIPTION
**Pull request by joelriou**
The pullback of sheaves was previously constructed as a composition of functors involving a left Kan extension. We redefine it as an abstract left adjoint of the pushforward functor when it exists, and in the most favourable case, we provide an isomorphism with the previous construction.
(This change shall be needed when we want to take pullback of étale sheaves.)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Topology/Sheaves/Functors.lean
Comment (kim-em) [inline Mathlib/Topology/Sheaves/Functors.lean:90]: Did you intend to delete this?

File: Mathlib/CategoryTheory/Sites/Pullback.lean
Comment (kim-em) [inline Mathlib/CategoryTheory/Sites/Pullback.lean:115]:
```suggestion
attribute [local instance] reflectsLimits_of_reflectsIsomorphisms in
```
(Safer if others add to this file later?)

File: Mathlib/Topology/Sheaves/Functors.lean
Comment (joelriou) [reply Mathlib/Topology/Sheaves/Functors.lean]: Yes, indeed, as it is no longer true, but the iso version is still there.

**Comment by joelriou** [comment]
Thanks for the reviews!

**Review by adamtopaz — APPROVED** [review]
LGTM! It would be great if @dagurtomas could also take a quick look before merging.

**Comment by dagurtomas** [comment]
Thanks!

maintainer merge

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**
